nixos/doc/manual/development/settings-options.xml

```diff
@@ -0,0 +1,169 @@
+           </para>
+         </listitem>
+       </varlistentry>
+     </variablelist>
+
+   </para>
+   <para xml:id="pkgs-formats-result">
+     These functions all return an attribute set with these values:
+    <variablelist>
+      <varlistentry>
+        <term>
+          <varname>type</varname>
+        </term>
+        <listitem>
+          <para>
+            A module system type representing a JSON value
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term>
+          <varname>generate</varname> <replaceable>filename</replaceable> <replaceable>jsonValue</replaceable>
+        </term>
+        <listitem>
+          <para>
+            A function that can render a JSON value to a file. Returns a file path.
+          </para>
+        </listitem>
+      </varlistentry>
+    </variablelist>
+   </para>
+   <example xml:id="ex-settings-nix-representable">
+     <title>Module with conventional <literal>settings</literal> option</title>
+     <para>
+       The following shows a module for an example program that uses a JSON configuration file. It demonstrates how above values can be used, along with some other related best practices. See the comments for explanations.
+     </para>
+<programlisting>
+{ options, config, lib, pkgs, ... }:
+let
+  cfg = config.services.foo;
+  # Define the settings format used for this program
+  settingsFormat = pkgs.formats.json {};
+in {
+
+  options.services.foo = {
+    enable = lib.mkEnableOption "foo service";
+
+    settings = lib.mkOption {
+      # Setting this type allows for correct merging behavior
+      type = settingsFormat.type;
+      default = {};
+      description = ''
+        Configuration for foo, see &lt;link xlink:href="https://example.com/docs/foo"/&gt;
+        for supported values.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    # We can assign some default settings here to make the service work by just enabling it.
+    # We use `mkDefault` for values that can be changed without problems
+    services.foo.settings = {
+      # Fails at runtime without any value set
+      log_level = lib.mkDefault "WARN";
+
+      # We assume systemd's `StateDirectory` is used, so we require this value, therefore no mkDefault
+      data_path = "/var/lib/foo";
+
+      # We use this to open the firewall, so we need to know about the default at eval time
+      port = lib.mkDefault 2546;
+    };
+
+    environment.etc."foo.json".source =
+      # The formats generator function takes a filename and the Nix value representing the format value
+      # and produces a filepath with that value rendered in the format
+      settingsFormat.generate "foo-config.json" cfg.settings;
+
+    # We know that `port` is assigned because we have a default value above
+    # So we can use it here without worries
+    networking.firewall.allowedTCPPorts = [ cfg.settings.port ];
```

Comment:
Maybe we should avoid showing adding port to firewall whitelist since modules are not supposed to do that.

Reply:
Oh good idea! I changed it to a `user` option now.